src/Spectre.libDataset.Tests/DatasetTest.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+* DatasetTest.cpp
+* Tests of Dataset implementation.
+*
+Copyright 2017 Maciej Gamrat, Grzegorz Mrukwa
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#define GTEST_LANG_CXX11 1
+
+#include <gtest/gtest.h>
+#include "Dataset.h"
+
+using namespace Spectre::libDataset;
+using samples = std::vector<int>;
+using samples_metadata = std::vector<int>;
+using dataset_metadata = std::vector<int>;
+using data_set = Dataset<int, int, dataset_metadata>;
+
+// @gmrukwa: provides general initialization test & emptiness tests
+namespace
+{
+TEST(DatasetInitializationTest, should_initialize_correctly)
+{
+    auto data = samples{ 1, 2, 3 };
+    auto sampleMetadata = samples_metadata{ 4, 5, 6 };
+    auto datasetMetadata = dataset_metadata{7};
+
+    data_set dataset(data, sampleMetadata, datasetMetadata);
+}
+
+TEST(DatasetTest, knows_if_empty)
+{
+    samples empty_samples;
+    samples_metadata empty_samples_metadata;
+    dataset_metadata empty_dataset_metadata;
+    data_set dataset(empty_samples, empty_samples_metadata, empty_dataset_metadata);
+    EXPECT_TRUE(dataset.empty());
```

Comment:
`empty` has no overloads for `const` or non-`const` return type. This would make no difference in that case.